056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+
+#### The new quorum check
+The restart conditions include a new check for controllers to verify that restarting the Kafka node does not affect the quorum health.
+The proposed check is:
+- Create admin client connection to the brokers and call `describeMetadataQuorum` API
+- If failed to connect to the brokers, return `UnforceableProblem` which currently results in delay and retry for the pod until the maximum attempts is reached.
+- From the quorum info returned from the admin API, read `lastCaughtUpTimestamp` of each controller. `lastCaughtUpTimestamp` is the last millisecond timestamp at which a replica controller was known to be caught up with the quorum leader.
+- Check the quorum leader id using the quorum info and identify the `lastCaughtUpTimestamp` of the quorum leader
+- Retrieve value of the Kafka property `controller.quorum.fetch.timeout.ms` from the desired configurations specified in the Kafka CR. If this property does not exist in the desired configurations, then use the hard-coded default value for it which is `2000`. The reason for this is explained further in the **NOTE** below.
+- Mark a Kafka node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`. This will exclude the controller KafkaRoller is currently considering to restart, because we are trying to check whether the quorum would stay healthy when this controller is restarted.
+- Count each controller node that is caught up, including the leader (`numOfCaughtUpControllers`).
+- Can restart if: `numOfCaughtUpControllers >= ceil((double) totalNumOfControllers / 2)`.
+
+> NOTE: Until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) is implemented, KafkaRoller cannot create an admin connection to the controller directly to describe its configuration or the quorum state. Therefore KafkaRoller checks the desired configurations to get the value of `controller.quorum.fetch.timeout.ms` and creates admin client connection to the brokers for the quorum check. If KafkaRoller cannot connect to any of the brokers after the maximum attempts to retry, the controllers will be marked as failed to reconcile because of not being able to determine the quorum health. KafkaRoller would then try to reconcile the brokers, which may help restoring admin client connections to them. In this scenario, the reconcilation will be completed with failure and reported to the operator. The controllers quorum check will be retried in the next round of reconcilation.
+
+
+#### Separate brokers and controllers
+For controller-only:
+ 1. Does not restart a controller performing log recovery
+ 2. Restarts a controller if the pod is in stuck state
+ 3. Attempts to connect an admin client to any of the brokers, if cannot connect, return `UnforceableProblem` which currently results in delay and retry for the pod until the maximum attempts is reached.
```

Comment:
If the max retry attempt is reached, what will we do in this case? 
I know you mentioned it in the above section, but it'd be good to add more description here to make it complete. ex:
` Attempts to connect an admin client to any of the brokers, if cannot connect, return UnforceableProblem which currently results in delay and retry for the pod until the maximum attempts is reached. Then trying to reconcile the brokers to hope to restore admin client connections issue.`